### PR TITLE
feat: add ray tracing initial support

### PR DESCRIPTION
**Pull request by manorit2001**

### CODE REVIEW

File: include/CIL/RayTracing/Camera.hpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    class Camera
+    {
+      public:
+        Camera(const double viewport_width, const double viewport_height,
+               const double focal_length = 1.0,
+               const Point3D origin = Point3D(0, 0, 0))
+            : m_origin(origin), m_horizontal(Vector3D(viewport_width, 0, 0)),
+              m_vertical(Vector3D(0.0, viewport_height, 0.0))
+        {
+            m_origin = Point3D(0, 0, 0);
```
Comment (parth-07) [inline include/CIL/RayTracing/Camera.hpp:17]: It seems like this line was added initially for testing, and now should be removed.

File: include/CIL/RayTracing/Camera.hpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        Ray get_ray(double u, double v) const
+        {
+            return Ray(m_origin, m_lower_left_corner + u * m_horizontal +
+                                     v * m_vertical - m_origin);
+        }
+
+        Ray get_ray(const ImageMatrix& img_data, const Pixel p) const
+        {
+            double u = static_cast<double>(p.row()) / img_data.width();
+            double v = static_cast<double>(p.col()) / img_data.height();
```
Comment (parth-07) [inline include/CIL/RayTracing/Camera.hpp:32]:
```suggestion
            double u = static_cast<double>(p.row()) / (img_data.width() - 1);
            double v = static_cast<double>(p.col()) / (img_data.height() - 1);
```

To account for row and col being 0-index based and width and height being 1-index based.

File: include/CIL/RayTracing/Camera.hpp (hunk 1 above)
Comment (manorit2001) [reply include/CIL/RayTracing/Camera.hpp:17]: ah yes, leftover refactor

File: include/CIL/RayTracing/Camera.hpp (hunk 2 above)
Comment (manorit2001) [reply include/CIL/RayTracing/Camera.hpp:32]: Makes sense

Comment (parth-07) [reply include/CIL/RayTracing/Camera.hpp:32]: There were a few more issues here, I have fixed them as well. Please review the updated code.

**Review by parth-07 — APPROVED** [review]
LGTM.

**Pull request merged**